src/loader/mod.rs

```diff
@@ -0,0 +1,578 @@
+        // the kernel is not lower than high memory.
+        if (lowest_kernel_start.is_some())
+            && (u64::from(boot_header.code32_start) < lowest_kernel_start.unwrap().raw_value())
+        {
+            return Err(Error::InvalidKernelStartAddress);
+        }
+
+        let mem_offset = match kernel_start {
+            Some(start) => start,
+            None => GuestAddress(u64::from(boot_header.code32_start)),
+        };
+
+        boot_header.code32_start = mem_offset.raw_value() as u32;
+
+        let mut loader_result: KernelLoaderResult = Default::default();
+        loader_result.setup_header = Some(boot_header);
+        loader_result.kernel_load = mem_offset;
+
+        //seek the compressed vmlinux.bin and read to memory
+        kernel_image
+            .seek(SeekFrom::Start(setup_size as u64))
+            .map_err(|_| Error::SeekBzImageCompressedKernel)?;
+        guest_mem
+            .read_exact_from(mem_offset, kernel_image, kernel_size)
+            .map_err(|_| Error::ReadBzImageCompressedKernel)?;
+
+        loader_result.kernel_end = mem_offset.raw_value() as GuestUsize + kernel_size as GuestUsize;
+
+        Ok(loader_result)
+    }
+}
+
+/// Writes the command line string to the given memory slice.
+///
+/// # Arguments
+///
+/// * `guest_mem` - A u8 slice that will be partially overwritten by the command line.
+/// * `guest_addr` - The address in `guest_mem` at which to load the command line.
+/// * `cmdline` - The kernel command line.
+pub fn load_cmdline(
+    guest_mem: &GuestMemoryMmap,
+    guest_addr: GuestAddress,
+    cmdline: &CStr,
+) -> Result<()> {
+    let len = cmdline.to_bytes().len();
+    if len == 0 {
+        return Ok(());
+    }
+
+    let end = guest_addr
+        .checked_add(len as u64 + 1)
+        .ok_or(Error::CommandLineOverflow)?; // Extra for null termination.
+    if end > guest_mem.end_addr() {
+        return Err(Error::CommandLineOverflow)?;
+    }
+
+    guest_mem
+        .write_slice(cmdline.to_bytes_with_nul(), guest_addr)
+        .map_err(|_| Error::CommandLineCopy)?;
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use std::io::Cursor;
+    use vm_memory::{Address, GuestAddress, GuestMemoryMmap};
+
+    const MEM_SIZE: u64 = 0x1000000;
+
+    fn create_guest_mem() -> GuestMemoryMmap {
+        GuestMemoryMmap::new(&[(GuestAddress(0x0), (MEM_SIZE as usize))]).unwrap()
+    }
+
+    #[allow(non_snake_case)]
+    fn make_bzImage() -> Vec<u8> {
+        let mut v = Vec::new();
+        v.extend_from_slice(include_bytes!("bzImage"));
+        v
+    }
+
+    // Elf64 image that prints hello world on x86_64.
+    fn make_elf_bin() -> Vec<u8> {
+        let mut v = Vec::new();
+        v.extend_from_slice(include_bytes!("test_elf.bin"));
+        v
+    }
+
+    #[allow(safe_packed_borrows)]
+    #[allow(non_snake_case)]
+    #[test]
+    #[ignore]
+    fn load_bzImage() {
+        let gm = create_guest_mem();
+        let image = make_bzImage();
```

Comment:
Even if you specify ignore, the test will still be compiled. Because of that compilation of tests fails with:
```
error: couldn't read src/loader/bzImage: No such file or directory (os error 2)
   --> src/loader/mod.rs:369:29
    |
369 |         v.extend_from_slice(include_bytes!("bzImage"));
    |                             ^^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error

error: Could not compile `linux-loader`.
```

I created an empty bzImage file to get rid of the error, but once we enable the CI it will fail there as well as the bzImage file doesn't exist. Right now I can't think of a nice way to go around this problem. There would be two not so nice options:
- add an empty bzImage file to the repository (I don't really like this one)
- make bzImage support a feature not enabled by default. This means that the tests will not be compiled by default either.

In the long run though I would like bzImage support to be tested as part of the CI as well to make sure we aren't introducing any bugs there. From a chat with @sameo I understood that we don't want to have a bzimage in the repository because of the license. How is this different from the elf image? Also, does the license apply to test code that is not part of the release binaries? It sounds kinda strange to me.

Reply:
@andreeaflorescu a bzImage would be compiled from the Linux kernel which is GPL-2 and so we would have to ship all the corresponding source / make an offer to provide the source.

The CI script for this repository could either fetch a prebuilt binary from an external source or could build a Linux kernel from scratch.

Reply:
I was thinking about providing an already built bzImage image like we are doing with elf. I don't see the need to always compile it from the linux source just so we can use it in the unit tests. For extensive testing with different kernel versions  we might want to result to integration testing instead.

Reply:
> I was thinking about providing an already built bzImage image like we are doing with elf. I don't see the need to always compile it from the linux source just so we can use it in the unit tests. For extensive testing with different kernel versions we might want to result to integration testing instead.

Unless we create something that looks like the bzImage, with it's custom binary layout, which would be a lot of work, the only straightforward way to get a bzImage is to build a Linux kernel. Shipping that built kernel in the repository would mean satisfying the GPL requirements which would be a large burden.

Reply:
@rbradford @andreeaflorescu Just want to confirm, to fix this issue, bzImage should be added to the server running CI test for rust-vmm crate, and specified under .buildkite/pipeline.yml as Rob mentioned. if my understanding is correct, will Andrea help put it or should I do?